x-pack/plugins/security_solution/public/flyout/shared/hooks/use_fetch_prevalence.ts

```diff
@@ -0,0 +1,180 @@
+  highlightedFields,
+  interval: { from, to },
+}: UseFetchPrevalenceParams): UseFetchPrevalenceResult => {
+  const {
+    services: {
+      data: { search: searchService },
+    },
+  } = useKibana();
+
+  const searchRequest = buildSearchRequest(highlightedFields, from, to);
+
+  const { data, isLoading, isError } = useQuery([QUERY_KEY, highlightedFields, from, to], () =>
+    createFetchData<RawAggregatedDataResponse>(searchService, searchRequest)
+  );
+
+  return {
+    loading: isLoading,
+    error: isError,
+    data,
+  };
+};
+
+/**
+ * Build the search request for the field/values pair, for a date range from/to.
+ * The request contains aggregation by aggregationField.
+ */
+const buildSearchRequest = (
+  highlightedFields: Record<string, QueryDslQueryContainer>,
+  from: string,
+  to: string
+): IEsSearchRequest => {
+  const query = buildEsQuery(
+    undefined,
+    [],
+    [
+      {
+        query: {
+          bool: {
+            filter: [
+              {
+                range: {
+                  '@timestamp': {
+                    gte: from,
+                    lte: to,
+                  },
+                },
+              },
+            ],
+          },
+        },
+        meta: {},
+      },
+    ]
+  );
+
+  return buildAggregationSearchRequest(query, highlightedFields);
+};
+
+const buildAggregationSearchRequest = (
+  query: QueryDslQueryContainer,
+  aggregationFilters: Record<string, QueryDslQueryContainer>
+): IEsSearchRequest => ({
+  params: {
+    body: {
+      query,
+      aggs: {
+        // with this aggregation, we can in a single call retrieve all the values for each field/value pairs
+        [FIELD_NAMES_AGG_KEY]: {
+          filters: {
+            filters: aggregationFilters,
+          },
+          aggs: {
+            // this sub aggregation allows us to retrieve all the hosts which have the field/value pair
+            [HOST_NAME_AGG_KEY]: {
+              terms: {
+                field: 'host.name',
+                size: 10000, // there could be a lot of hosts which have the field/value pair
+              },
+            },
+            // this sub aggregation allows us to retrieve all the users which have the field/value pair
+            [USER_NAME_AGG_KEY]: {
+              terms: {
+                field: 'user.name',
+                size: 10000, // there could be a lot of users which have the field/value pair
+              },
+            },
+            // we use this sub aggregation to differenciate between alerts (event.kind === 'signal') and documents (event.kind !== 'signal')
+            [EVENT_KIND_AGG_KEY]: {
+              terms: {
+                field: 'event.kind',
+                size: 10, // there should be only 8 different value for the event.kind field
+              },
+            },
+          },
+        },
+        // retrieve all the unique hosts in the environment
+        [HOSTS_AGG_KEY]: {
+          terms: {
```

Comment:
Looking at the actual response, it returns an array of host and their records, i think this could be a performance issue down the line for big customers, whenever they open the flyout it will fetch up to 10000 hosts and 10000 users. 

have you checked out using cardinality to pull unique host and users https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html

Reply:
you are brilliant! ❤️